libvast/src/format/csv.cpp

```diff
@@ -129,4 +140,271 @@ const char* writer::name() const {
+}
+
+caf::optional<record_type>
+reader::make_layout(const std::vector<std::string>& names) {
+  for (auto& t : schema_) {
+    if (auto r = caf::get_if<record_type>(&t)) {
+      auto select_fields = [&]() -> caf::optional<record_type> {
+        std::vector<record_field> result_raw;
+        for (auto& name : names) {
+          if (auto field = r->at(name))
+            result_raw.emplace_back(name, *field);
+          else
+            return caf::none;
+        }
+        return record_type{result_raw};
+      };
+      if (auto result = select_fields())
+        return result->name(r->name());
+    } else if (names.size() == 1 && names[0] == t.name()) {
+      // Hoist naked type into record.
+      // TODO: Maybe this is actually not a good idea?
+      return record_type{{t.name(), t}}.name(t.name());
+    } // else skip
+  }
+  return caf::none;
+}
+
+namespace {
+
+template <class Iterator, class Attribute>
+struct container_parser_builder {
+  using result_type = rule<Iterator, Attribute>;
+
+  explicit container_parser_builder(const std::string& set_separator)
+    : set_separator_{set_separator} {
+    // nop
+  }
+
+  // TODO: support enumeration_type inside of containers too.
+  template <class T>
+  result_type operator()(const T& t) const {
+    if constexpr (std::is_same_v<T, string_type>) {
+      return +(parsers::any - set_separator_ - '=')->*[](std::string&& x) {
+        return data{std::move(x)};
+      };
+    } else if constexpr (std::is_same_v<T, pattern_type>) {
+      return +(parsers::any - set_separator_ - '=')->*[](std::string&& x) {
+        return data{pattern{std::move(x)}};
+      };
+    } else if constexpr (std::is_same_v<T, set_type>) {
+      auto set_insert = [](std::vector<Attribute>&& xs) {
+        return set(std::make_move_iterator(xs.begin()),
+                   std::make_move_iterator(xs.end()));
+      };
+      return ('{'_p >> (caf::visit(*this, t.value_type) % set_separator_)
+              >> '}'_p)
+               ->*set_insert;
+    } else if constexpr (std::is_same_v<T, vector_type>) {
+      auto vector_insert = [](std::vector<Attribute>&& xs) { return xs; };
+      return ('[' >> (caf::visit(*this, t.value_type) % set_separator_) >> ']')
+               ->*vector_insert;
+    } else if constexpr (std::is_same_v<T, map_type>) {
+      auto map_insert = [](std::vector<std::tuple<Attribute, Attribute>>&& xs) {
+        auto to_pair = [](auto&& tuple) {
+          return std::make_pair(std::get<0>(tuple), std::get<1>(tuple));
+        };
+        map m;
+        for (auto& x : xs)
+          m.insert(to_pair(std::move(x)));
+        return m;
+      };
+      auto kvp = caf::visit(*this, t.key_type) >> '='
+                 >> caf::visit(*this, t.value_type);
+      return (('{'_p >> (kvp % set_separator_) >> '}')->*map_insert);
+    } else if constexpr (has_parser_v<type_to_data<T>>) {
+      using value_type = type_to_data<T>;
+      auto ws = ignore(*parsers::space);
```

Comment:
So are we skipping whitespace already? Probably that's the location that we want to have configurable with the `csv_parser_options` struct.

Reply:
The string-like types are not handled in this path.